src/v/datalake/translation/state_machine.cc

```diff
@@ -0,0 +1,147 @@
+    auto val = datalake::translation::translation_state{
+      .highest_translated_offset = offset};
+    storage::record_batch_builder builder(
+      model::record_batch_type::datalake_translation_state, model::offset(0));
+    builder.add_raw_kv(std::nullopt, serde::to_iobuf(val));
+    auto batch = std::move(builder).build();
+    return model::make_memory_record_batch_reader(std::move(batch));
+}
+
+} // namespace
+
+namespace datalake::translation {
+
+translation_stm::translation_stm(ss::logger& logger, raft::consensus* raft)
+  : raft::persisted_stm<>("datalake_translation_stm.snapshot", logger, raft) {}
+
+ss::future<> translation_stm::do_apply(const model::record_batch& batch) {
+    if (
+      batch.header().type
+      != model::record_batch_type::datalake_translation_state) {
+        co_return;
+    }
+    auto record_iterator = model::record_batch_iterator::create(batch);
+    while (record_iterator.has_next()) {
+        auto record = record_iterator.next();
+        auto value = serde::from_iobuf<translation_state>(
+          record.release_value());
+        vlog(
+          _log.trace,
+          "updating highest translated offset to {}",
+          value.highest_translated_offset);
+        _highest_translated_offset = std::max(
+          _highest_translated_offset, value.highest_translated_offset);
+    }
+}
+
+ss::future<std::optional<kafka::offset>>
+translation_stm::highest_translated_offset(
+  model::timeout_clock::duration timeout) {
+    if (!_raft->log_config().iceberg_enabled() || !co_await sync(timeout)) {
+        co_return std::nullopt;
+    }
+    co_return _highest_translated_offset;
+}
+
+ss::future<std::error_code> translation_stm::reset_highest_translated_offset(
+  kafka::offset new_translated_offset,
+  model::term_id term,
+  model::timeout_clock::duration timeout,
+  ss::abort_source& as) {
+    if (!co_await sync(timeout) || _insync_term != term) {
+        co_return raft::errc::not_leader;
+    }
+    vlog(
+      _log.debug,
+      "Reset-ing highest translated offset to {} from {} in term: {}",
+      new_translated_offset,
+      _highest_translated_offset,
+      term);
+    auto current_term = _insync_term;
+    // We are at a newer or equal term than the entry, so likely the
+    // stm has gotten out of sync
+    if (_highest_translated_offset >= new_translated_offset) {
+        co_return raft::errc::success;
+    }
+    auto result = co_await _raft->replicate(
+      current_term,
+      make_translation_state_batch(new_translated_offset),
+      make_replicate_options());
+    auto deadline = model::timeout_clock::now() + timeout;
+    if (
+      result
+      && co_await wait_no_throw(
+        result.value().last_offset, deadline, std::ref(as))) {
+        co_return raft::errc::success;
+    }
+    if (!as.abort_requested() && _raft->term() == current_term) {
+        co_await _raft->step_down("datalake coordinator sync error");
```

Comment:
if this check fails (lets assume wait_no_throw() timed out for some hypothetical reason due to a delay in the applier, but raft is up to date), the state machine is in an inconsistent state from the log, meaning there are entries that are raft committed but not applied in the state machine.

So for a class of state machine implementations (which the code calls ingestion time stms), that look up the state machine more applying a change they operate on the inconsistent state. In this case a subsequent reset_highest_translated_offset() can potentially perform an incorrect check in L86, so this step down resets the state (as subsequent sync fails). Normally this is accompanied by a lock so this whole state change in  the lock scope but we have only one reader/writer for this stm.

Reply:
I see, thanks for the context. I guess generally this helps avoid surprise updates to the STM when the decisions that build the updates are based on stale data. OTOH, it seems like this is also "handled" by making the STM robust to stale updates with the std::max() call (this is also somewhat what I'm doing in the coordinator, which applies a rough CAS)

No need to change anything, but just wondering if the std::max() in apply is enough now

Reply:
right.. I think coordinator replicates first and reconciles later which the "ingestion time stm" tries to avoid as in some cases (eg: produce path) it is expensive to replicate, rather it resorts to tricks like this to ensure the checks always happen on the latest snapshot of data.